# Data Visualisation

This notebook aims to explore the data retrived from ENCODE.

In [1]:
# loading the data
import pandas as pd

metadata_df = pd.read_csv('HepG2_data/HepG2_DNAm/metadata.tsv', sep='\t')
print(metadata_df.head())


  File accession   File format File type File format type  \
0    ENCFF690FNR     bed bed9+       bed            bed9+   
1    ENCFF884UAE        bigWig    bigWig              NaN   
2    ENCFF464CWC        bigWig    bigWig              NaN   
3    ENCFF549SNG  bigBed bed9+    bigBed            bed9+   
4    ENCFF840XVU     bed bed9+       bed            bed9+   

                             Output type File assembly Experiment accession  \
0               methylation state at CpG        GRCh38          ENCSR786DCL   
1   plus strand methylation state at CpG        GRCh38          ENCSR786DCL   
2  minus strand methylation state at CpG        GRCh38          ENCSR786DCL   
3               methylation state at CpG        GRCh38          ENCSR786DCL   
4               methylation state at CpG        GRCh38          ENCSR625HZA   

  Assay                    Donor(s) Biosample term id  ... Platform  \
0  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   
1  WGBS  /human-

WGBS - methylation data

In [10]:
WGBS_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF857YRG.bed', sep='\t', nrows=10, header =None)
print(WGBS_df)

     0      1      2   3   4  5      6      7        8   9    10  11  12  13
0  chr1  10468  10469   1   1  +  10468  10469  0,255,0   1    0  CG  CN   1
1  chr1  10496  10497   1   1  +  10496  10497  255,0,0   1  100  CG  CG   2
2  chr1  10524  10525   1   1  +  10524  10525  255,0,0   1  100  CG  CG   2
3  chr1  10541  10542   1   1  +  10541  10542  255,0,0   1  100  CG  CG   2
4  chr1  10562  10563   1   1  +  10562  10563  255,0,0   1  100  CG  CG   2
5  chr1  10570  10571   1   1  +  10570  10571  255,0,0   1  100  CG  CG   2
6  chr1  10576  10577   1   1  +  10576  10577  255,0,0   1  100  CG  CG   2
7  chr1  10578  10579   1   1  +  10578  10579  255,0,0   1  100  CG  CG   2
8  chr1  10588  10589   1   1  +  10588  10589  255,0,0   1  100  CG  CG   2
9  chr1  10608  10609   1   1  +  10608  10609  255,0,0   1  100  CG  CG   2


In [19]:
# loading WGBS data (methylation)

WGBS_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF690FNR.bed', sep='\t', names=['refChr', 'startChr', 'endChr', 'name', 'score', 'strand', 'startThick', 'endThick', 'rgb', 'reads', 'pctMe', 'refGen', 'sampGen', 'qualScore'], header =None)
WGBS_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF884UAE.bedGraph', sep='\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)
WGBS_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF464CWC.bedGraph', sep='\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)

MemoryError: Unable to allocate 512. KiB for an array with shape (65536,) and data type int64

Columns in big bed files (in FNR bed file):
- Reference chromosome or scaffold
- Start position in chromosome
- End position in chromosome
- Name of item
- Score from 0-1000. Capped number of reads
- Strandedness, plus (+), minus (-), or unknown (?)
- Start of where display should be thick
- End of where display should be thick
- Color value (RGB)
- Coverage, or number of reads
- Percentage of reads that show methylation at this position in the genome
- Reference genotype
- Sample genotype
- Quality score for genotype call


Final column in bigwig files = percent methylation at CpG site

In [3]:
display(WGBS_df.head())
display(WGBS_plus_df.head())
display(WGBS_minus_df.head())

refChr  startChr  endChr  name  score strand  startThick  endThick  \
0   chr1     10468   10469     1      1      +       10468     10469   
1   chr1     10469   10470     1      4      -       10469     10470   
2   chr1     10470   10471     1      1      +       10470     10471   
3   chr1     10471   10472     1      2      -       10471     10472   
4   chr1     10483   10484     1      1      +       10483     10484   

         rgb  reads  pctMe refGen sampGen  qualScore  
0    0,255,0      1      0     CG      MG          3  
1  155,255,0      4     25     CG      CK          5  
2    255,0,0      1    100     CG      CG         11  
3    255,0,0      2    100     CG      CG          6  
4    255,0,0      1    100     CG      CG          4

refChr  start    end  pctMe
0   chr1  10009  10012    0.0
1   chr1  10015  10018    0.0
2   chr1  10021  10024    0.0
3   chr1  10027  10030    0.0
4   chr1  10033  10036    0.0

refChr  start    end  pctMe
0   chr1  10469  10470   25.0
1   chr1  10471  10472  100.0
2   chr1  10472  10473    0.0
3   chr1  10481  10482    0.0
4   chr1  10484  10485  100.0

RNA data - plus/minus strand and gene quant data

In [3]:
#loading RNA data

RNA_geneQuant_df = pd.read_csv('HepG2_data/HepG2_DNAm/ENCFF649XOG.tsv', sep='\t', header=0)
RNA_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF194TSC.bedGraph', sep='\t', names=['refChr', 'start', 'end', 'exp'])
RNA_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF792BNC.bedGraph', sep='\t', names=['refChr', 'start', 'end', 'exp'])

Columns for RNA:

- chromosome
- start
- end
- gene expression?

In [4]:
display(RNA_plus_df)
display(RNA_minus_df)
display(RNA_geneQuant_df)

refChr     start       end      exp
0         chr1    190711    190761  0.02116
1         chr1    190786    190808  0.02116
2         chr1    190808    190836  0.06349
3         chr1    190836    190858  0.04233
4         chr1    190864    190892  0.04233
...        ...       ...       ...      ...
7198639   chrY  22823877  22823927  0.02116
7198640   chrY  22824904  22824954  0.02116
7198641   chrY  22825018  22825067  0.02116
7198642   chrY  56871389  56871439  0.02116
7198643   chrY  56871467  56871517  0.02116

[7198644 rows x 4 columns]

refChr     start       end      exp
0         chr1     14475     14490  0.02116
1         chr1     14490     14525  0.04233
2         chr1     14525     14540  0.02116
3         chr1     14660     14668  0.02116
4         chr1     14668     14675  0.04233
...        ...       ...       ...      ...
7063920   chrY  24213384  24213405  0.02116
7063921   chrY  24213792  24213821  0.02116
7063922   chrY  26621781  26621831  0.02116
7063923   chrY  26621878  26621895  0.02116
7063924   chrY  26622511  26622544  0.02116

[7063925 rows x 4 columns]

gene_id     transcript_id(s)  length  effective_length  \
0                    10904                10904    93.0              0.00   
1                    12954                12954    94.0              0.00   
2                    12956                12956    72.0              0.00   
3                    12958                12958    82.0              0.00   
4                    12960                12960    73.0              0.00   
...                    ...                  ...     ...               ...   
59521  gSpikein_ERCC-00165  tSpikein_ERCC-00165   872.0            700.78   
59522  gSpikein_ERCC-00168  tSpikein_ERCC-00168  1024.0            852.78   
59523  gSpikein_ERCC-00170  tSpikein_ERCC-00170  1023.0            851.78   
59524  gSpikein_ERCC-00171  tSpikein_ERCC-00171   505.0            333.85   
59525     gSpikein_phiX174     tSpikein_phiX174  5386.0           5214.78   

       expected_count     TPM     FPKM  posterior_mean_count  \
0                 0.0    0.00     0.00                   0.0   
1                 0.0    0.00     0.00                   0.0   
2                 0.0    0.00     0.00                   0.0   
3                 0.0    0.00     0.00                   0.0   
4                 0.0    0.00     0.00                   0.0   
...               ...     ...      ...                   ...   
59521            87.0    4.47     8.43                  87.0   
59522             1.0    0.04     0.08                   1.0   
59523            51.0    2.16     4.07                  51.0   
59524          3942.0  425.45   802.16                3942.0   
59525         78337.0  541.27  1020.55               78337.0   

       posterior_standard_deviation_of_count  pme_TPM  pme_FPKM  \
0                                        0.0     0.00      0.00   
1                                        0.0     0.00      0.00   
2                                        0.0     0.00      0.00   
3                                        0.0     0.00      0.00   
4                                        0.0     0.00      0.00   
...                                      ...      ...       ...   
59521                                    0.0     4.33      8.42   
59522                                    0.0     0.08      0.16   
59523                                    0.0     2.10      4.09   
59524                                    0.0   406.87    791.53   
59525                                    0.0   517.51   1006.78   

       TPM_ci_lower_bound  TPM_ci_upper_bound  \
0                0.000000            0.000000   
1                0.000000            0.000000   
2                0.000000            0.000000   
3                0.000000            0.000000   
4                0.000000            0.000000   
...                   ...                 ...   
59521            3.439680            5.244820   
59522            0.002566            0.194005   
59523            1.548640            2.687830   
59524          394.250000          419.834000   
59525          513.704000          521.421000   

       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \
0                                   0.000000             0.000000   
1                                   0.000000             0.000000   
2                                   0.000000             0.000000   
3                                   0.000000             0.000000   
4                                   0.000000             0.000000   
...                                      ...                  ...   
59521                               0.071704             6.672670   
59522                               0.474234             0.005171   
59523                               0.093629             3.011470   
59524                               0.010739           767.089000   
59525                               0.002592           999.854000   

       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  
0                 0.000000       

In [5]:
RNA_geneQuant_df[10000:11000]

gene_id                                   transcript_id(s)  \
10000   ENSG00000151006.7  ENST00000280606.6,ENST00000486499.1,ENST000004...   
10001  ENSG00000151012.13                ENST00000280612.9,ENST00000509248.1   
10002   ENSG00000151014.5  ENST00000280614.3,ENST00000515616.1,ENST000006...   
10003  ENSG00000151023.16  ENST00000331161.8,ENST00000376363.5,ENST000004...   
10004  ENSG00000151025.10  ENST00000376351.3,ENST00000482641.2,ENST000004...   
...                   ...                                                ...   
10995   ENSG00000159387.7                ENST00000290552.7,ENST00000558315.1   
10996   ENSG00000159388.5                ENST00000290551.4,ENST00000475157.1   
10997  ENSG00000159398.15  ENST00000290567.13,ENST00000319165.13,ENST0000...   
10998   ENSG00000159399.9  ENST00000290573.6,ENST00000409174.1,ENST000004...   
10999  ENSG00000159403.16  ENST00000535233.6,ENST00000536053.6,ENST000005...   

        length  effective_length  expected_count   TPM   FPKM  \
10000  4510.90           4339.68          124.04  1.03   1.94   
10001  9249.21           9077.99         1848.00  7.33  13.83   
10002  1856.43           1685.21          115.00  2.46   4.64   
10003  1318.00           1146.78            2.00  0.06   0.12   
10004  4853.25           4682.04            0.00  0.00   0.00   
...        ...               ...             ...   ...    ...   
10995  1989.00           1817.78            0.00  0.00   0.00   
10996  2579.06           2407.84          286.00  4.28   8.07   
10997  1766.22           1595.01            0.00  0.00   0.00   
10998  5250.08           5078.88          920.95  6.53  12.32   
10999  2664.70           2493.49           84.00  1.21   2.29   

       posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \
10000                134.12                                   6.64     1.21   
10001               1848.00                                   0.00     7.14   
10002                115.00                                   0.00     2.63   
10003                  2.00                                   0.00     0.21   
10004                  0.00                                   0.00     0.08   
...                     ...                                    ...      ...   
10995                  0.00                                   0.00     0.06   
10996                286.00                                   0.00     4.15   
10997                  0.00                                   0.00     0.25   
10998                920.91                                   0.29     6.44   
10999                 83.63                                   0.66     2.60   

       pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \
10000      2.36            0.872066            1.566580   
10001     13.88            6.625750            7.708410   
10002      5.11            2.013440            3.298030   
10003      0.41            0.048159            0.425282   
10004      0.15            0.004778            0.197461   
...         ...                 ...                 ...   
10995      0.12            0.000902            0.167404   
10996      8.07            3.643900            4.658430   
10997      0.50            0.085282            0.469493   
10998     12.54            5.750730            7.220230   
10999      5.05            1.830850            3.407680   

       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \
10000                               0.102375             1.704870   
10001                               0.025499            12.886400   
10002                               0.083487             3.915400   
10003                               0.317992             0.093792   
10004                               0.489654             0.010009   
...                                      ...                  ...   
10995                               0.532145             0.001434   
10996                               0.042490             7.082570 

Seems to be a combination of ensemble genes, spike-ins (calibrations) and tRNA (digits - i.e. 10904)

Histone ChIP-Seq data - H3K9me3 and HeK27me3 modifications

In [8]:
# loading histone mod data
# H3K9me3
histone_pval_K9_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF125NHB.bedGraph', sep='\t', names=['chrRef', 'start', 'end', 'pval'])
histone_peaks_K9_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF170GMN.bed', sep='\t', header=None)

# H3K27me3
histone_pval_K27_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF896BFP.bedGraph', sep='\t', names=['chrRef', 'start', 'end', 'pval'])
histone_peaks_K27_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF841QVP.bed', sep='\t', header=None)


Histone columns:
- Chromosome 
- start
- end
- peak label?
- 

peaks:
- chromosome
- start
- end 
- replication pct? 

In [ ]:
display(histone_peaks_K9_df)
display(histone_pval_K9_df)

display(histone_peaks_K27_df)
display(histone_pval_K27_df)


0         1         2            3     4  5        6          7  \
0       chr1    788438    790667   Peak_33943   206  .  4.85493   20.69931   
1       chr1    991782    992368   Peak_49693   167  .  5.23056   16.71326   
2       chr1   1112984   1113962   Peak_13060   297  .  7.35106   29.75177   
3       chr1   1194302   1195375    Peak_6971   362  .  8.27128   36.28568   
4       chr1   1770030   1770624  Peak_107498    98  .  3.79372    9.87359   
...      ...       ...       ...          ...   ... ..      ...        ...   
148025  chrY  26667979  26668865   Peak_48031   172  .  4.20870   17.27292   
148026  chrY  26668976  26671930       Peak_4  1000  .  6.71736  200.44463   
148027  chrY  26672410  26672736   Peak_91129   112  .  3.60158   11.25807   
148028  chrY  56679663  56680112   Peak_60182   151  .  4.94783   15.15114   
148029  chrY  56683855  56684306   Peak_10462   321  .  6.92330   32.11600   

                8     9  
0        18.43085   635  
1        14.64974   235  
2        27.03791   471  
3        33.31710   650  
4         8.21870   315  
...           ...   ...  
148025   15.19134   673  
148026  194.12437  2116  
148027    9.52160    94  
148028   13.17521   124  
148029   29.30741   227  

[148030 rows x 10 columns]

chrRef  start    end     pval
0                            chr1      0   9944  0.00100
1                            chr1   9944   9947  0.02630
2                            chr1   9947   9950  0.13984
3                            chr1   9950   9954  0.39194
4                            chr1   9954  10150  0.79457
...                           ...    ...    ...      ...
251263757  chrY_KI270740v1_random  17991  33769  0.00100
251263758  chrY_KI270740v1_random  33769  33984  0.02630
251263759  chrY_KI270740v1_random  33984  34006  0.13984
251263760  chrY_KI270740v1_random  34006  34212  0.02630
251263761  chrY_KI270740v1_random  34212  34224  0.00100

[251263762 rows x 4 columns]

0          1          2           3   4  5        6        7  \
0      chr1  100236060  100236175  Peak_91116  33  .  3.38068  3.39518   
1      chr1  100538099  100538408   Peak_9577  65  .  5.03472  6.59658   
2      chr1  101135095  101135210  Peak_97001  32  .  3.29575  3.26597   
3      chr1  101236811  101237102  Peak_17149  61  .  4.77464  6.16004   
4      chr1  101237336  101237500  Peak_87875  36  .  3.27881  3.62443   
...     ...        ...        ...         ...  .. ..      ...      ...   
51278  chrX     848527     848652  Peak_65447  43  .  3.93203  4.39984   
51279  chrX    8742048    8742163  Peak_65448  43  .  3.93203  4.39984   
51280  chrX     915821     915961   Peak_5605  77  .  6.13741  7.77377   
51281  chrX    9295740    9295902  Peak_31774  54  .  4.48337  5.46977   
51282  chrX     966389     966573  Peak_48390  49  .  4.47639  4.98931   

             8    9  
0      0.84457   88  
1      2.72930   79  
2      0.83925   47  
3      2.57973  107  
4      1.03439   55  
...        ...  ...  
51278  1.48232   36  
51279  1.48232   26  
51280  3.56466   85  
51281  2.10106   87  
51282  1.86485  111  

[51283 rows x 10 columns]

chrRef  start    end     pval
0                           chr1      0  10071  0.30592
1                           chr1  10071  10100  0.82513
2                           chr1  10100  10127  0.50508
3                           chr1  10127  10186  0.18005
4                           chr1  10186  10215  0.04769
...                          ...    ...    ...      ...
59656086  chrY_KI270740v1_random  29725  31121  0.30592
59656087  chrY_KI270740v1_random  31121  31236  0.82513
59656088  chrY_KI270740v1_random  31236  36460  0.30592
59656089  chrY_KI270740v1_random  36460  36487  0.82513
59656090  chrY_KI270740v1_random  36487  36575  0.50508

[59656091 rows x 4 columns]

## Visualising

To begin, we utilised the HepG2 WGBS methylation state at CPG, RNA-Seq Gene Quantifications, and Histone ChIP-Seq pseudoreplicated peak files (loading them in their original forms). 
- ENCFF690FNR (first 9 columns only): WGBS
- ENCFF194TSC: RNA minus
- ENCFF792BNC: RNA plus
- ENCFF170GMN: Hist H3K9me3 pseudorep
- ENCFF505LOU: Hist H3K27me3 pseudorep

WGBS

RNA

In [ ]:
# combining RNA gene quantification file with V29 gene annotations?

Histone ChIP-Seq

## Intervals
First, annotate data with genes and aggregate at the gene level.

Compute the sum of modifications for an entire gene plus its transcription start site. An example model might have the signature: (sum DNA methylation, sum H3K9me3, ...etc) -> (gene expression level). 

In [ ]:
# Gene annotations for WBGS data - can be applied directly to it using chr, start and end
# Gene annotations for RNA data - not sure how to approach this
# Gene annotations for hist CS data - do I used pseudo or p-val? annotate both using bedtools intersect

In [4]:
display(WGBS_df)
WGBS_df.iloc[1:10,0:15]

refChr  startChr  endChr  name  score strand  \
0                           chr1     10468   10469     1      1      +   
1                           chr1     10469   10470     1      4      -   
2                           chr1     10470   10471     1      1      +   
3                           chr1     10471   10472     1      2      -   
4                           chr1     10483   10484     1      1      +   
...                          ...       ...     ...   ...    ...    ...   
51350653  chrY_KI270740v1_random     35645   35646     1      1      +   
51350654  chrY_KI270740v1_random     35646   35647     1      1      -   
51350655  chrY_KI270740v1_random     35685   35686     1      1      +   
51350656  chrY_KI270740v1_random     35686   35687     1      1      -   
51350657  chrY_KI270740v1_random     35862   35863     1      3      -   

          startThick  endThick        rgb  reads  pctMe refGen sampGen  \
0              10468     10469    0,255,0      1      0     CG      MG   
1              10469     10470  155,255,0      4     25     CG      CK   
2              10470     10471    255,0,0      1    100     CG      CG   
3              10471     10472    255,0,0      2    100     CG      CG   
4              10483     10484    255,0,0      1    100     CG      CG   
...              ...       ...        ...    ...    ...    ...     ...   
51350653       35645     35646    0,255,0      1      0     CG      CG   
51350654       35646     35647    0,255,0      1      0     CG      CG   
51350655       35685     35686    0,255,0      1      0     CG      CG   
51350656       35686     35687    0,255,0      1      0     CG      CG   
51350657       35862     35863    255,0,0      3    100     CG      CG   

          qualScore  
0                 3  
1                 5  
2                11  
3                 6  
4                 4  
...             ...  
51350653          3  
51350654          3  
51350655          3  
51350656          3  
51350657          6  

[51350658 rows x 14 columns]

refChr  startChr  endChr  name  score strand  startThick  endThick  \
1   chr1     10469   10470     1      4      -       10469     10470   
2   chr1     10470   10471     1      1      +       10470     10471   
3   chr1     10471   10472     1      2      -       10471     10472   
4   chr1     10483   10484     1      1      +       10483     10484   
5   chr1     10484   10485     1      1      -       10484     10485   
6   chr1     10488   10489     1      1      +       10488     10489   
7   chr1     10489   10490     1      1      -       10489     10490   
8   chr1     10492   10493     1      1      +       10492     10493   
9   chr1     10493   10494     1      1      -       10493     10494   

         rgb  reads  pctMe refGen sampGen  qualScore  
1  155,255,0      4     25     CG      CK          5  
2    255,0,0      1    100     CG      CG         11  
3    255,0,0      2    100     CG      CG          6  
4    255,0,0      1    100     CG      CG          4  
5    255,0,0      1    100     CG      CG          4  
6    255,0,0      1    100     CG      CG          4  
7    255,0,0      1    100     CG      CG          4  
8    255,0,0      1    100     CG      MG          3  
9    0,255,0      1      0     CG      CK          3

In [3]:
#loading gene annotated wgbs data
WGBS_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF884UAE.bedGraph', sep='\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)
WGBS_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF464CWC.bedGraph', sep='\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)



ParserError: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.

In [21]:
display(wgbs_annotated_df)
display(wgbs_annotated_df.iloc[0:10, 0:15])
#display(wgbs_annotated_df2.iloc[0:10])


0       1       2   3   4  5       6       7          8   9   ...  13  \
0     chr1   14709   14710   1   1  +   14709   14710    255,0,0   1  ...   2   
1     chr1   14709   14710   1   1  +   14709   14710    255,0,0   1  ...   2   
2     chr1   14715   14716   1   1  +   14715   14716    255,0,0   1  ...   2   
3     chr1   14715   14716   1   1  +   14715   14716    255,0,0   1  ...   2   
4     chr1   14740   14741   1   1  +   14740   14741    255,0,0   1  ...   2   
...    ...     ...     ...  ..  .. ..     ...     ...        ...  ..  ...  ..   
9995  chr1  805670  805671   1  12  -  805670  805671   255,55,0  12  ...  24   
9996  chr1  805670  805671   1  12  -  805670  805671   255,55,0  12  ...  24   
9997  chr1  805670  805671   1  12  -  805670  805671   255,55,0  12  ...  24   
9998  chr1  805723  805724   1   6  -  805723  805724  105,255,0   6  ...   3   
9999  chr1  805723  805724   1   6  -  805723  805724  105,255,0   6  ...   3   

        14      15          16      17      18 19  20  21  \
0     chr1  HAVANA        gene   14404   29570  .   -   .   
1     chr1  HAVANA  transcript   14404   29570  .   -   .   
2     chr1  HAVANA        gene   14404   29570  .   -   .   
3     chr1  HAVANA  transcript   14404   29570  .   -   .   
4     chr1  HAVANA        gene   14404   29570  .   -   .   
...    ...     ...         ...     ...     ... ..  ..  ..   
9995  chr1  HAVANA        gene  586071  827796  .   -   .   
9996  chr1  HAVANA  transcript  594308  827796  .   -   .   
9997  chr1  HAVANA        gene  778770  810060  .   +   .   
9998  chr1  HAVANA  transcript  804799  807465  .   +   .   
9999  chr1  HAVANA  transcript  804932  810060  .   +   .   

                                                     22  
0     gene_id "ENSG00000227232.5"; gene_type "unproc...  
1     gene_id "ENSG00000227232.5"; transcript_id "EN...  
2     gene_id "ENSG00000227232.5"; gene_type "unproc...  
3     gene_id "ENSG00000227232.5"; transcript_id "EN...  
4     gene_id "ENSG00000227232.5"; gene_type "unproc...  
...                                                 ...  
9995  gene_id "ENSG00000230021.10"; gene_type "trans...  
9996  gene_id "ENSG00000230021.10"; transcript_id "E...  
9997  gene_id "ENSG00000237491.8"; gene_type "lincRN...  
9998  gene_id "ENSG00000237491.8"; transcript_id "EN...  
9999  gene_id "ENSG00000237491.8"; transcript_id "EN...  

[10000 rows x 23 columns]

0      1      2   3   4  5      6      7        8   9    10  11  12  13  \
0  chr1  14709  14710   1   1  +  14709  14710  255,0,0   1  100  CG  CG   2   
1  chr1  14709  14710   1   1  +  14709  14710  255,0,0   1  100  CG  CG   2   
2  chr1  14715  14716   1   1  +  14715  14716  255,0,0   1  100  CG  CG   2   
3  chr1  14715  14716   1   1  +  14715  14716  255,0,0   1  100  CG  CG   2   
4  chr1  14740  14741   1   1  +  14740  14741  255,0,0   1  100  CG  CG   2   
5  chr1  14740  14741   1   1  +  14740  14741  255,0,0   1  100  CG  CG   2   
6  chr1  14746  14747   1   1  +  14746  14747  255,0,0   1  100  CG  CG   2   
7  chr1  14746  14747   1   1  +  14746  14747  255,0,0   1  100  CG  CG   2   
8  chr1  14772  14773   1   1  +  14772  14773  255,0,0   1  100  CG  CG   2   
9  chr1  14772  14773   1   1  +  14772  14773  255,0,0   1  100  CG  CG   2   

     14  
0  chr1  
1  chr1  
2  chr1  
3  chr1  
4  chr1  
5  chr1  
6  chr1  
7  chr1  
8  chr1  
9  chr1

0      1      2   3   4  5      6      7          8   9   ...  13 14 15  \
0  chr1  10468  10469   1   1  +  10468  10469    0,255,0   1  ...   3  .  .   
1  chr1  10469  10470   1   4  -  10469  10470  155,255,0   4  ...   5  .  .   
2  chr1  10470  10471   1   1  +  10470  10471    255,0,0   1  ...  11  .  .   
3  chr1  10471  10472   1   2  -  10471  10472    255,0,0   2  ...   6  .  .   
4  chr1  10483  10484   1   1  +  10483  10484    255,0,0   1  ...   4  .  .   
5  chr1  10484  10485   1   1  -  10484  10485    255,0,0   1  ...   4  .  .   
6  chr1  10488  10489   1   1  +  10488  10489    255,0,0   1  ...   4  .  .   
7  chr1  10489  10490   1   1  -  10489  10490    255,0,0   1  ...   4  .  .   
8  chr1  10492  10493   1   1  +  10492  10493    255,0,0   1  ...   3  .  .   
9  chr1  10493  10494   1   1  -  10493  10494    0,255,0   1  ...   3  .  .   

   16 17 18 19  20  21 22  
0   . -1 -1  .   .   .  .  
1   . -1 -1  .   .   .  .  
2   . -1 -1  .   .   .  .  
3   . -1 -1  .   .   .  .  
4   . -1 -1  .   .   .  .  
5   . -1 -1  .   .   .  .  
6   . -1 -1  .   .   .  .  
7   . -1 -1  .   .   .  .  
8   . -1 -1  .   .   .  .  
9   . -1 -1  .   .   .  .  

[10 rows x 23 columns]

In [8]:
wgbs_annotated_df = pd.read_csv('HepG2_data/HepG2_DNAm/annotated/wgbs_with_gene_annotations_loj.bed', sep='\t', header=None, usecols=[0, 1, 2, 5, 10, 22] )
display(wgbs_annotated_df.iloc[0:10,0:15])

0      1      2  5    10 22
0  chr1  10468  10469  +    0  .
1  chr1  10469  10470  -   25  .
2  chr1  10470  10471  +  100  .
3  chr1  10471  10472  -  100  .
4  chr1  10483  10484  +  100  .
5  chr1  10484  10485  -  100  .
6  chr1  10488  10489  +  100  .
7  chr1  10489  10490  -  100  .
8  chr1  10492  10493  +  100  .
9  chr1  10493  10494  -    0  .

In [4]:
pattern = r'gene_id\s*"([^"]+)"'

wgbs_annotated_df.loc[:, 'GeneID'] = wgbs_annotated_df.iloc[:, -1].str.extract(pattern)
wgbs_annotated_df.loc[:, 'GeneID'] = wgbs_annotated_df.iloc[:, -1].str.extract(pattern)

In [11]:
test = wgbs_annotated_df.iloc[0:1]
display(test)
test.loc[:, 'GeneID'] = test.iloc[:, -1].str.extract(pattern)
display(test)
test2 = wgbs_annotated_df.iloc[0:1]
print(test2[22])

0      1      2  5   10 22
0  chr1  10468  10469  +   0  .

C:\Users\julia\AppData\Local\Temp\ipykernel_9980\2183389383.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test.loc[:, 'GeneID'] = test.iloc[:, -1].str.extract(pattern)


0      1      2  5  10 22 GeneID
0  chr1  10468  10469  +   0  .    NaN

0    .
Name: 22, dtype: object


In [5]:
display(wgbs_annotated_df)

0      1      2  5   10 22 GeneID
0                            chr1  10468  10469  +    0  .    NaN
1                            chr1  10469  10470  -   25  .    NaN
2                            chr1  10470  10471  +  100  .    NaN
3                            chr1  10471  10472  -  100  .    NaN
4                            chr1  10483  10484  +  100  .    NaN
...                           ...    ...    ... ..  ... ..    ...
156345308  chrY_KI270740v1_random  35645  35646  +    0  .    NaN
156345309  chrY_KI270740v1_random  35646  35647  -    0  .    NaN
156345310  chrY_KI270740v1_random  35685  35686  +    0  .    NaN
156345311  chrY_KI270740v1_random  35686  35687  -    0  .    NaN
156345312  chrY_KI270740v1_random  35862  35863  -  100  .    NaN

[156345313 rows x 7 columns]

In [4]:
# annotate histone chipseq files
h3k9me3_pp_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\t', header=None, nrows=1000000)
#h3k9me3_pval_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\t', header=None, nrows=1000000)
h3k27me3_pp_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\t', header=None, nrows=1000000)
#h3k27me3_pval_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\t', header=None, nrows=1000000)

In [9]:
# checking that expected count hasn't been altered

display(histone_peaks_K9_df[0:10])
display(histone_peaks_K27_df[0:10])
display(h3k9me3_pp_annotated_df[0:10])
display(h3k27me3_pp_annotated_df[0:10])

0        1        2            3    4  5        6         7         8  \
0  chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   
1  chr1   991782   992368   Peak_49693  167  .  5.23056  16.71326  14.64974   
2  chr1  1112984  1113962   Peak_13060  297  .  7.35106  29.75177  27.03791   
3  chr1  1194302  1195375    Peak_6971  362  .  8.27128  36.28568  33.31710   
4  chr1  1770030  1770624  Peak_107498   98  .  3.79372   9.87359   8.21870   
5  chr1  1771153  1772290    Peak_9246  334  .  6.34875  33.41285  30.55444   
6  chr1  2077026  2078865    Peak_2373  493  .  9.31909  49.38892  45.98370   
7  chr1  2422970  2423938   Peak_19118  260  .  6.78560  26.06830  23.51803   
8  chr1  2429237  2430460   Peak_45478  175  .  5.37193  17.51166  15.40157   
9  chr1  2488544  2489340   Peak_92748  109  .  3.80781  10.96670   9.23946   

      9  
0   635  
1   235  
2   471  
3   650  
4   315  
5   225  
6  1237  
7   570  
8   859  
9   232

0          1          2           3   4  5        6        7        8  \
0  chr1  100236060  100236175  Peak_91116  33  .  3.38068  3.39518  0.84457   
1  chr1  100538099  100538408   Peak_9577  65  .  5.03472  6.59658  2.72930   
2  chr1  101135095  101135210  Peak_97001  32  .  3.29575  3.26597  0.83925   
3  chr1  101236811  101237102  Peak_17149  61  .  4.77464  6.16004  2.57973   
4  chr1  101237336  101237500  Peak_87875  36  .  3.27881  3.62443  1.03439   
5  chr1  101240137  101240569  Peak_71693  40  .  3.59275  4.07966  1.32727   
6  chr1  101250307  101250422  Peak_71694  40  .  3.59275  4.07966  1.32727   
7  chr1  101314378  101314496  Peak_39327  50  .  4.22304  5.04547  1.91208   
8  chr1   10695837   10696061  Peak_34048  52  .  4.37074  5.28382  2.07027   
9  chr1    1070212    1070504   Peak_4701  78  .  5.93909  7.81863  3.59440   

     9  
0   88  
1   79  
2   47  
3  107  
4   55  
5   94  
6   84  
7   64  
8  174  
9   70

0        1        2           3    4  5        6         7         8   \
0  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
1  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
2  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
3  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
4  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   
5  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   
6  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   
7  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   
8  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   
9  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   

    9     10       11          12       13       14 15 16 17  \
0  635  chr1   HAVANA        gene   586071   827796  .  -  .   
1  635  chr1   HAVANA  transcript   594308   827796  .  -  .   
2  635  chr1   HAVANA        gene   778770   810060  .  +  .   
3  635  chr1   HAVANA  transcript   779056   804875  .  +  .   
4  471  chr1   HAVANA        gene  1081818  1116361  .  -  .   
5  471  chr1   HAVANA  transcript  1081818  1116081  .  -  .   
6  471  chr1   HAVANA  transcript  1081823  1116098  .  -  .   
7  650  chr1   HAVANA        gene  1173884  1197935  .  +  .   
8  650  chr1   HAVANA  transcript  1173903  1197935  .  +  .   
9  650  chr1  ENSEMBL  transcript  1173906  1197933  .  +  .   

                                                  18  
0  gene_id "ENSG00000230021.10"; gene_type "trans...  
1  gene_id "ENSG00000230021.10"; transcript_id "E...  
2  gene_id "ENSG00000237491.8"; gene_type "lincRN...  
3  gene_id "ENSG00000237491.8"; transcript_id "EN...  
4  gene_id "ENSG00000131591.17"; gene_type "prote...  
5  gene_id "ENSG00000131591.17"; transcript_id "E...  
6  gene_id "ENSG00000131591.17"; transcript_id "E...  
7  gene_id "ENSG00000162571.13"; gene_type "prote...  
8  gene_id "ENSG00000162571.13"; transcript_id "E...  
9  gene_id "ENSG00000162571.13"; transcript_id "E...

0        1        2           3    4  5        6         7         8   \
0  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
1  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
2  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
3  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   
4  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   
5  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   
6  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   
7  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   
8  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   
9  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   

    9     10       11          12       13       14 15 16 17  \
0  635  chr1   HAVANA        gene   586071   827796  .  -  .   
1  635  chr1   HAVANA  transcript   594308   827796  .  -  .   
2  635  chr1   HAVANA        gene   778770   810060  .  +  .   
3  635  chr1   HAVANA  transcript   779056   804875  .  +  .   
4  471  chr1   HAVANA        gene  1081818  1116361  .  -  .   
5  471  chr1   HAVANA  transcript  1081818  1116081  .  -  .   
6  471  chr1   HAVANA  transcript  1081823  1116098  .  -  .   
7  650  chr1   HAVANA        gene  1173884  1197935  .  +  .   
8  650  chr1   HAVANA  transcript  1173903  1197935  .  +  .   
9  650  chr1  ENSEMBL  transcript  1173906  1197933  .  +  .   

                                                  18  
0  gene_id "ENSG00000230021.10"; gene_type "trans...  
1  gene_id "ENSG00000230021.10"; transcript_id "E...  
2  gene_id "ENSG00000237491.8"; gene_type "lincRN...  
3  gene_id "ENSG00000237491.8"; transcript_id "EN...  
4  gene_id "ENSG00000131591.17"; gene_type "prote...  
5  gene_id "ENSG00000131591.17"; transcript_id "E...  
6  gene_id "ENSG00000131591.17"; transcript_id "E...  
7  gene_id "ENSG00000162571.13"; gene_type "prote...  
8  gene_id "ENSG00000162571.13"; transcript_id "E...  
9  gene_id "ENSG00000162571.13"; transcript_id "E...

In [43]:
display(h3k27me3_pp_annotated_df)
display(h3k9me3_pp_annotated_df)

0         1         2            3    4  5        6         7   \
0       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
1       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
2       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
3       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
4       chr1   1112984   1113962   Peak_13060  297  .  7.35106  29.75177   
...      ...       ...       ...          ...  ... ..      ...       ...   
210848  chrY  26618172  26618975  Peak_136047   81  .  3.53416   8.12781   
210849  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   
210850  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   
210851  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   
210852  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   

              8    9     10      11          12        13        14 15 16 17  \
0       18.43085  635  chr1  HAVANA        gene    586071    827796  .  -  .   
1       18.43085  635  chr1  HAVANA  transcript    594308    827796  .  -  .   
2       18.43085  635  chr1  HAVANA        gene    778770    810060  .  +  .   
3       18.43085  635  chr1  HAVANA  transcript    779056    804875  .  +  .   
4       27.03791  471  chr1  HAVANA        gene   1081818   1116361  .  -  .   
...          ...  ...   ...     ...         ...       ...       ... .. .. ..   
210848   6.56647  650  chrY  HAVANA  transcript  26594851  26634652  .  -  .   
210849   5.98529  162  chrY  HAVANA        gene  26594851  26634652  .  -  .   
210850   5.98529  162  chrY  HAVANA  transcript  26594851  26634652  .  -  .   
210851   3.84296  493  chrY  HAVANA        gene  26594851  26634652  .  -  .   
210852   3.84296  493  chrY  HAVANA  transcript  26594851  26634652  .  -  .   

                                                       18  
0       gene_id "ENSG00000230021.10"; gene_type "trans...  
1       gene_id "ENSG00000230021.10"; transcript_id "E...  
2       gene_id "ENSG00000237491.8"; gene_type "lincRN...  
3       gene_id "ENSG00000237491.8"; transcript_id "EN...  
4       gene_id "ENSG00000131591.17"; gene_type "prote...  
...                                                   ...  
210848  gene_id "ENSG00000237917.1"; transcript_id "EN...  
210849  gene_id "ENSG00000237917.1"; gene_type "unproc...  
210850  gene_id "ENSG00000237917.1"; transcript_id "EN...  
210851  gene_id "ENSG00000237917.1"; gene_type "unproc...  
210852  gene_id "ENSG00000237917.1"; transcript_id "EN...  

[210853 rows x 19 columns]

0         1         2            3    4  5        6         7   \
0       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
1       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
2       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
3       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   
4       chr1   1112984   1113962   Peak_13060  297  .  7.35106  29.75177   
...      ...       ...       ...          ...  ... ..      ...       ...   
210848  chrY  26618172  26618975  Peak_136047   81  .  3.53416   8.12781   
210849  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   
210850  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   
210851  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   
210852  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   

              8    9     10      11          12        13        14 15 16 17  \
0       18.43085  635  chr1  HAVANA        gene    586071    827796  .  -  .   
1       18.43085  635  chr1  HAVANA  transcript    594308    827796  .  -  .   
2       18.43085  635  chr1  HAVANA        gene    778770    810060  .  +  .   
3       18.43085  635  chr1  HAVANA  transcript    779056    804875  .  +  .   
4       27.03791  471  chr1  HAVANA        gene   1081818   1116361  .  -  .   
...          ...  ...   ...     ...         ...       ...       ... .. .. ..   
210848   6.56647  650  chrY  HAVANA  transcript  26594851  26634652  .  -  .   
210849   5.98529  162  chrY  HAVANA        gene  26594851  26634652  .  -  .   
210850   5.98529  162  chrY  HAVANA  transcript  26594851  26634652  .  -  .   
210851   3.84296  493  chrY  HAVANA        gene  26594851  26634652  .  -  .   
210852   3.84296  493  chrY  HAVANA  transcript  26594851  26634652  .  -  .   

                                                       18  
0       gene_id "ENSG00000230021.10"; gene_type "trans...  
1       gene_id "ENSG00000230021.10"; transcript_id "E...  
2       gene_id "ENSG00000237491.8"; gene_type "lincRN...  
3       gene_id "ENSG00000237491.8"; transcript_id "EN...  
4       gene_id "ENSG00000131591.17"; gene_type "prote...  
...                                                   ...  
210848  gene_id "ENSG00000237917.1"; transcript_id "EN...  
210849  gene_id "ENSG00000237917.1"; gene_type "unproc...  
210850  gene_id "ENSG00000237917.1"; transcript_id "EN...  
210851  gene_id "ENSG00000237917.1"; gene_type "unproc...  
210852  gene_id "ENSG00000237917.1"; transcript_id "EN...  

[210853 rows x 19 columns]

## Intervals

aggregating to the gene level for WGBS and histone pp data

In [45]:
#filtering to keep only "gene" rows (removing "transcript", "exon", etc.)
wgbs_annotated_gene_df = wgbs_annotated_df[wgbs_annotated_df[16] == 'gene']
h3k9me3_pp_annotated_gene_df = h3k9me3_pp_annotated_df[h3k9me3_pp_annotated_df[12] == 'gene']
h3k27me3_pp_annotated_gene_df = h3k27me3_pp_annotated_df[h3k27me3_pp_annotated_df[12] == 'gene']

print(h3k9me3_pp_annotated_gene_df[18][0])
display(wgbs_annotated_gene_df.head())
display(h3k9me3_pp_annotated_gene_df.head())
display(h3k27me3_pp_annotated_gene_df.head())

gene_id "ENSG00000230021.10"; gene_type "transcribed_processed_pseudogene"; gene_name "AL669831.3"; level 2; havana_gene "OTTHUMG00000191652.4";


0      1      2   3   4  5      6      7        8   9   ...  13    14  \
0  chr1  14709  14710   1   1  +  14709  14710  255,0,0   1  ...   2  chr1   
2  chr1  14715  14716   1   1  +  14715  14716  255,0,0   1  ...   2  chr1   
4  chr1  14740  14741   1   1  +  14740  14741  255,0,0   1  ...   2  chr1   
6  chr1  14746  14747   1   1  +  14746  14747  255,0,0   1  ...   2  chr1   
8  chr1  14772  14773   1   1  +  14772  14773  255,0,0   1  ...   2  chr1   

       15    16     17     18 19  20  21  \
0  HAVANA  gene  14404  29570  .   -   .   
2  HAVANA  gene  14404  29570  .   -   .   
4  HAVANA  gene  14404  29570  .   -   .   
6  HAVANA  gene  14404  29570  .   -   .   
8  HAVANA  gene  14404  29570  .   -   .   

                                                  22  
0  gene_id "ENSG00000227232.5"; gene_type "unproc...  
2  gene_id "ENSG00000227232.5"; gene_type "unproc...  
4  gene_id "ENSG00000227232.5"; gene_type "unproc...  
6  gene_id "ENSG00000227232.5"; gene_type "unproc...  
8  gene_id "ENSG00000227232.5"; gene_type "unproc...  

[5 rows x 23 columns]

0        1        2            3    4  5        6         7         8   \
0   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   
2   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   
4   chr1  1112984  1113962   Peak_13060  297  .  7.35106  29.75177  27.03791   
7   chr1  1194302  1195375    Peak_6971  362  .  8.27128  36.28568  33.31710   
10  chr1  1770030  1770624  Peak_107498   98  .  3.79372   9.87359   8.21870   

     9     10      11    12       13       14 15 16 17  \
0   635  chr1  HAVANA  gene   586071   827796  .  -  .   
2   635  chr1  HAVANA  gene   778770   810060  .  +  .   
4   471  chr1  HAVANA  gene  1081818  1116361  .  -  .   
7   650  chr1  HAVANA  gene  1173884  1197935  .  +  .   
10  315  chr1  HAVANA  gene  1751232  1780457  .  -  .   

                                                   18  
0   gene_id "ENSG00000230021.10"; gene_type "trans...  
2   gene_id "ENSG00000237491.8"; gene_type "lincRN...  
4   gene_id "ENSG00000131591.17"; gene_type "prote...  
7   gene_id "ENSG00000162571.13"; gene_type "prote...  
10  gene_id "ENSG00000008130.15"; gene_type "prote...

0        1        2            3    4  5        6         7         8   \
0   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   
2   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   
4   chr1  1112984  1113962   Peak_13060  297  .  7.35106  29.75177  27.03791   
7   chr1  1194302  1195375    Peak_6971  362  .  8.27128  36.28568  33.31710   
10  chr1  1770030  1770624  Peak_107498   98  .  3.79372   9.87359   8.21870   

     9     10      11    12       13       14 15 16 17  \
0   635  chr1  HAVANA  gene   586071   827796  .  -  .   
2   635  chr1  HAVANA  gene   778770   810060  .  +  .   
4   471  chr1  HAVANA  gene  1081818  1116361  .  -  .   
7   650  chr1  HAVANA  gene  1173884  1197935  .  +  .   
10  315  chr1  HAVANA  gene  1751232  1780457  .  -  .   

                                                   18  
0   gene_id "ENSG00000230021.10"; gene_type "trans...  
2   gene_id "ENSG00000237491.8"; gene_type "lincRN...  
4   gene_id "ENSG00000131591.17"; gene_type "prote...  
7   gene_id "ENSG00000162571.13"; gene_type "prote...  
10  gene_id "ENSG00000008130.15"; gene_type "prote...

Extracting gene IDs from annotated dataframes

In [6]:
# extracting gene ID
h3k9me3_pp_annotated_gene_df = h3k9me3_pp_annotated_df[h3k9me3_pp_annotated_df[12] == 'gene']
h3k27me3_pp_annotated_gene_df = h3k27me3_pp_annotated_df[h3k27me3_pp_annotated_df[12] == 'gene']
pattern = r'gene_id\s*"([^"]+)"'
h3k9me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k9me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)
h3k27me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k27me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)
#wgbs_annotated_gene_df.loc[:, 'GeneID'] = wgbs_annotated_gene_df.iloc[:, -1].str.extract(pattern)

C:\Users\julia\AppData\Local\Temp\ipykernel_15684\3790687905.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  h3k9me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k9me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)
C:\Users\julia\AppData\Local\Temp\ipykernel_15684\3790687905.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  h3k27me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k27me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)


Aggregating data by gene

WGBS aggregating by gene. 1st attempt = aggregating methylation percentage


do I only aggregate CpG-rich regions? How to approach this? filter by coverage? Ensure reliable results?

In [47]:
# Aggregating WGBS data by gene and average pct methylation per bp in gene (avg pct per gene bp)
wgbs_annotated_gene_df

0        1        2  3  4  5        6        7        8  9  ...  \
0       chr1    14709    14710  1  1  +    14709    14710  255,0,0  1  ...   
2       chr1    14715    14716  1  1  +    14715    14716  255,0,0  1  ...   
4       chr1    14740    14741  1  1  +    14740    14741  255,0,0  1  ...   
6       chr1    14746    14747  1  1  +    14746    14747  255,0,0  1  ...   
8       chr1    14772    14773  1  1  +    14772    14773  255,0,0  1  ...   
...      ...      ...      ... .. .. ..      ...      ...      ... ..  ...   
999957  chr1  9722512  9722513  1  1  +  9722512  9722513  255,0,0  1  ...   
999963  chr1  9722513  9722514  1  3  -  9722513  9722514  255,0,0  3  ...   
999969  chr1  9722516  9722517  1  2  +  9722516  9722517  255,0,0  2  ...   
999975  chr1  9722517  9722518  1  3  -  9722517  9722518  255,0,0  3  ...   
999991  chr1  9722532  9722533  1  2  +  9722532  9722533  255,0,0  2  ...   

          14      15    16       17       18 19 20  21  \
0       chr1  HAVANA  gene    14404    29570  .  -   .   
2       chr1  HAVANA  gene    14404    29570  .  -   .   
4       chr1  HAVANA  gene    14404    29570  .  -   .   
6       chr1  HAVANA  gene    14404    29570  .  -   .   
8       chr1  HAVANA  gene    14404    29570  .  -   .   
...      ...     ...   ...      ...      ... .. ..  ..   
999957  chr1  HAVANA  gene  9651732  9729114  .  +   .   
999963  chr1  HAVANA  gene  9651732  9729114  .  +   .   
999969  chr1  HAVANA  gene  9651732  9729114  .  +   .   
999975  chr1  HAVANA  gene  9651732  9729114  .  +   .   
999991  chr1  HAVANA  gene  9651732  9729114  .  +   .   

                                                       22              GeneID  
0       gene_id "ENSG00000227232.5"; gene_type "unproc...   ENSG00000227232.5  
2       gene_id "ENSG00000227232.5"; gene_type "unproc...   ENSG00000227232.5  
4       gene_id "ENSG00000227232.5"; gene_type "unproc...   ENSG00000227232.5  
6       gene_id "ENSG00000227232.5"; gene_type "unproc...   ENSG00000227232.5  
8       gene_id "ENSG00000227232.5"; gene_type "unproc...   ENSG00000227232.5  
...                                                   ...                 ...  
999957  gene_id "ENSG00000171608.15"; gene_type "prote...  ENSG00000171608.15  
999963  gene_id "ENSG00000171608.15"; gene_type "prote...  ENSG00000171608.15  
999969  gene_id "ENSG00000171608.15"; gene_type "prote...  ENSG00000171608.15  
999975  gene_id "ENSG00000171608.15"; gene_type "prote...  ENSG00000171608.15  
999991  gene_id "ENSG00000171608.15"; gene_type "prote...  ENSG00000171608.15  

[254680 rows x 24 columns]

In [64]:
wgbs_annotated_gene_df.iloc[1:10,0:15]

0      1      2  3  4  5      6      7         8  9   10  11  12  13  \
2   chr1  14715  14716  1  1  +  14715  14716   255,0,0  1  100  CG  CG   2   
4   chr1  14740  14741  1  1  +  14740  14741   255,0,0  1  100  CG  CG   2   
6   chr1  14746  14747  1  1  +  14746  14747   255,0,0  1  100  CG  CG   2   
8   chr1  14772  14773  1  1  +  14772  14773   255,0,0  1  100  CG  CG   2   
10  chr1  14774  14775  1  1  +  14774  14775   255,0,0  1  100  CG  CG   2   
12  chr1  14790  14791  1  1  +  14790  14791   255,0,0  1  100  CG  CG   2   
14  chr1  15643  15644  1  1  -  15643  15644   255,0,0  1  100  CG  CG   2   
16  chr1  16243  16244  1  1  -  16243  16244   255,0,0  1  100  CG  CG   2   
18  chr1  16554  16555  1  7  +  16554  16555  255,55,0  7   85  CG  CG  10   

      14  
2   chr1  
4   chr1  
6   chr1  
8   chr1  
10  chr1  
12  chr1  
14  chr1  
16  chr1  
18  chr1

In [51]:
## need to fix logic here
# get pct and get the average over each bp
wgbs_pct_avg_per_gene = wgbs_annotated_gene_df.groupby('GeneID').agg({10: "mean"}).reset_index()

In [56]:
display(wgbs_pct_avg_per_gene)
wgbs_pct_avg_per_gene[10].describe()

GeneID         10
0    ENSG00000007923.15  90.242105
1    ENSG00000008128.22  83.959079
2    ENSG00000008130.15  75.072961
3    ENSG00000041988.15  71.348031
4    ENSG00000049239.12  80.770075
..                  ...        ...
302   ENSG00000284745.1  28.580292
303   ENSG00000284747.1  92.269231
304   ENSG00000285629.1  46.465854
305   ENSG00000285812.1  93.048077
306   ENSG00000285945.1  16.409091

[307 rows x 2 columns]

count    307.000000
mean      57.631266
std       28.367165
min        0.000000
25%       35.248829
50%       59.333333
75%       83.562500
max      100.000000
Name: 10, dtype: float64

Histone peak aggregation

In [7]:
# Histone aggregating by gene
# sum peaks per gene? 
# what do the other columns represent?

k9_peaks_per_gene = h3k9me3_pp_annotated_gene_df.groupby('GeneID').size().reset_index(name='PeakCount')
k27_peaks_per_gene = h3k27me3_pp_annotated_gene_df.groupby('GeneID').size().reset_index(name='PeakCount')

display(k9_peaks_per_gene)
display(k27_peaks_per_gene)

GeneID  PeakCount
0     ENSG00000000457.13          1
1     ENSG00000000460.16          1
2     ENSG00000000971.15          6
3     ENSG00000001084.12          5
4      ENSG00000001629.9          3
...                  ...        ...
8451   ENSG00000285978.1          7
8452   ENSG00000285981.1          2
8453   ENSG00000285986.1          1
8454   ENSG00000285987.1          5
8455   ENSG00000285989.1          1

[8456 rows x 2 columns]

GeneID  PeakCount
0     ENSG00000000457.13          1
1     ENSG00000000460.16          1
2     ENSG00000000971.15          6
3     ENSG00000001084.12          5
4      ENSG00000001629.9          3
...                  ...        ...
8451   ENSG00000285978.1          7
8452   ENSG00000285981.1          2
8453   ENSG00000285986.1          1
8454   ENSG00000285987.1          5
8455   ENSG00000285989.1          1

[8456 rows x 2 columns]

RNA aggregation

RNA quant file contains ensemble genes
Annotate ensemble genes with corresponding V29 gene
A mapping file has been created to extract the chromosome, start site, end site, enseml name and V29 name
They will be merged below:

In [10]:
import pandas as pd

# loading the mapping file and gene quant file
rna_quant_annotated = pd.read_csv('HepG2_data/HepG2_DNAm/annotated/rna_gene_quant_with_annotations.csv', header=[0])
rna_quant_annotated_genes = rna_quant_annotated[rna_quant_annotated['gene_name'].notna()]

In [12]:
rna_quant_annotated_genes

gene_id                                   transcript_id(s)  \
649    ENSG00000000003.14  ENST00000373020.8,ENST00000494424.1,ENST000004...   
650     ENSG00000000005.5                ENST00000373031.4,ENST00000485971.1   
651    ENSG00000000419.12  ENST00000371582.8,ENST00000371584.8,ENST000003...   
652    ENSG00000000457.13  ENST00000367770.5,ENST00000367771.10,ENST00000...   
653    ENSG00000000460.16  ENST00000286031.10,ENST00000359326.8,ENST00000...   
...                   ...                                                ...   
59424   ENSG00000285990.1                                  ENST00000649331.1   
59425   ENSG00000285991.1                                  ENST00000647612.1   
59426   ENSG00000285992.1                                  ENST00000648949.1   
59427   ENSG00000285993.1                                  ENST00000650266.1   
59428   ENSG00000285994.1                                  ENST00000648650.1   

        length  effective_length  expected_count    TPM   FPKM  \
649    2179.14           2007.92         2019.00  36.23  68.31   
650     940.50            769.31            0.00   0.00   0.00   
651    1078.03            906.82          887.00  35.24  66.45   
652    3488.94           3317.73          258.00   2.80   5.28   
653    2911.75           2740.54          879.00  11.56  21.79   
...        ...               ...             ...    ...    ...   
59424   647.00            475.81            0.75   0.06   0.11   
59425  5065.00           4893.78           14.93   0.11   0.21   
59426   956.00            784.78            0.00   0.00   0.00   
59427  1246.00           1074.78            0.00   0.00   0.00   
59428  3732.00           3560.78            0.00   0.00   0.00   

       posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \
649                 2019.00                                   0.00    34.91   
650                    0.00                                   0.00     0.12   
651                  887.00                                   0.00    34.27   
652                  258.00                                   0.00     2.79   
653                  879.00                                   0.00    11.50   
...                     ...                                    ...      ...   
59424                  1.80                                   1.60     0.20   
59425                 17.18                                   5.91     0.13   
59426                  0.00                                   0.00     0.04   
59427                  0.00                                   0.00     0.03   
59428                  0.00                                   0.00     0.01   

       ...  TPM_ci_lower_bound  TPM_ci_upper_bound  \
649    ...        3.334520e+01           36.533100   
650    ...        1.760980e-03            0.313410   
651    ...        3.199170e+01           36.595700   
652    ...        2.313660e+00            3.287400   
653    ...        1.024230e+01           12.846600   
...    ...                 ...                 ...   
59424  ...        5.294150e-06            0.530734   
59425  ...        4.175390e-02            0.234562   
59426  ...        1.021180e-06            0.131430   
59427  ...        2.000310e-06            0.095424   
59428  ...        1.194670e-07            0.028976   

       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \
649                                 0.015839         6.478720e+01   
650                                 0.511119         3.288750e-03   
651                                 0.023183         6.222900e+01   
652                                 0.060300         4.505810e+00   
653                                 0.039000         1.991680e+01   
...                                      ...                  ...   
59424                               0.602857         1.030280e-05   
59425                               0.266380         7.841970e-02   
59426                               0.654743 

RNA is now annotated. We can group by gene and aggregate by expected count.

In [36]:
rna_count_per_gene = rna_quant_annotated_genes.groupby('gene_id').agg({'expected_count': "sum"}).reset_index()

In [37]:
display(rna_count_per_gene)
print(rna_count_per_gene['expected_count'].describe())

gene_id  expected_count
0      ENSG00000000003.14         2019.00
1       ENSG00000000005.5            0.00
2      ENSG00000000419.12          887.00
3      ENSG00000000457.13          258.00
4      ENSG00000000460.16          879.00
...                   ...             ...
58716   ENSG00000285990.1            0.75
58717   ENSG00000285991.1           14.93
58718   ENSG00000285992.1            0.00
58719   ENSG00000285993.1            0.00
58720   ENSG00000285994.1            0.00

[58721 rows x 2 columns]

count     58721.000000
mean        245.880392
std        3587.987114
min           0.000000
25%           0.000000
50%           0.000000
75%          11.000000
max      532366.000000
Name: expected_count, dtype: float64


## Normalisation

Normalisation will utilise 'light-house genes'... XXX

In [ ]:
# For .bigBed files
#bb = pybedtools.BedTool('HepG2_DNAm/ENCFF549SNG.bigBed')
#for feature in bb:
    # process feature, which is an interval in the BED file
    #print(feature)